### PR TITLE
Run operations sequentially in Threads::Task when n_threads==1.

### PR DESCRIPTION
**Pull request by bangerth**

### CODE REVIEW

File: include/deal.II/base/thread_management.h [hunk 1]
```diff
@@ -975,16 +1005,48 @@ namespace Threads
...
      * @post Using this constructor automatically makes the task object
      * joinable().
      */
     Task(const std::function<RT()> &function_object)
     {
-      task_data = std::make_shared<TaskData>(
-        std::move(std::async(std::launch::async, function_object)));
+      task_data = std::make_shared<TaskData>();
+
+      if (MultithreadInfo::n_threads() > 0)
```
Comment (tjhei) [inline include/deal.II/base/thread_management.h:1022]: do you mean > 1 here?

File: include/deal.II/base/thread_management.h [hunk 2]
```diff
@@ -1142,11 +1204,25 @@ namespace Threads
     class TaskData
     {
     public:
-      TaskData(std::future<RT> &&future)
-        : future(std::move(future))
-        , task_has_finished(false)
+      /**
+       *
```
Comment (tjhei) [inline include/deal.II/base/thread_management.h:1208]: documentation?

File: include/deal.II/base/thread_management.h [hunk 3]
```diff
@@ -1231,6 +1307,12 @@ namespace Threads
    * function object without arguments and returning an object of type RT (or
    * void).
    *
+   * @note When MultithreadInfo::n_threads() returns 1, i.e., if the
+   *   deal.II runtime system has been configured to only use one
+   *   thread, then this function just executes the given function
+   *   object and stores the return value in the Task object returned
```
Comment (tjhei) [inline include/deal.II/base/thread_management.h:1313]:
```suggestion
   *   object immediately and stores the return value in the Task object returned
```

File: include/deal.II/base/thread_management.h [hunk 4]
```diff
@@ -1276,6 +1358,12 @@ namespace Threads
    * can later retrieve via <code>task.return_value()</code> once the
    * task (i.e., the body of the lambda function) has completed.
    *
+   * @note When MultithreadInfo::n_threads() returns 1, i.e., if the
+   *   deal.II runtime system has been configured to only use one
+   *   thread, then this function just executes the given function
+   *   object and stores the return value in the Task object returned
```
Comment (tjhei) [inline include/deal.II/base/thread_management.h:1364]:
```suggestion
   *   object immediately and stores the return value in the Task object returned
```

**Comment by bangerth** [comment]
All addressed.

File: include/deal.II/base/thread_management.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/thread_management.h]: Yes, doh!

**Comment by tjhei** [comment]
```
An error occurred in line <2734> of file </jenkins/workspace/dealii_PR-10412/source/fe/mapping_q_generic.cc> in function

    dealii::CellSimilarity::Similarity dealii::MappingQGeneric<dim, spacedim>::fill_fe_values(const typename dealii::Triangulation<dim, spacedim>::cell_iterator&, dealii::CellSimilarity::Similarity, const dealii::Quadrature<dim>&, const typename dealii::Mapping<dim, spacedim>::InternalDataBase&, dealii::internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>&) const [with int dim = 2; int spacedim = 2; typename dealii::Triangulation<dim, spacedim>::cell_iterator = dealii::TriaIterator<dealii::CellAccessor<2, 2> >; typename dealii::Mapping<dim, spacedim>::InternalDataBase = dealii::Mapping<2, 2>::InternalDataBase]

The violated condition was: 

    dynamic_cast<const InternalData *>(&internal_data) != nullptr

Additional information: 

    This exception -- which is used in many places in the library -- usually indicates that some condition which the author of the code thought must be satisfied at a certain point in an algorithm, is not fulfilled. An example would be that the first part of an algorithm sorts elements of an array in ascending order, and a second part of the algorithm later encounters an element that is not larger than the previous one.



There is usually not very much you can do if you encounter such an exception since it indicates an error in deal.II, not in your own program. Try to come up with the smallest possible program that still demonstrates the error and contact the deal.II mailing lists with it to obtain help.



Stacktrace:

-----------

#0  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: dealii::MappingQGeneric<2, 2>::fill_fe_values(dealii::TriaIterator<dealii::CellAccessor<2, 2> > const&, dealii::CellSimilarity::Similarity, dealii::Quadrature<2> const&, dealii::Mapping<2, 2>::InternalDataBase const&, dealii::internal::FEValuesImplementation::MappingRelatedData<2, 2>&) const

#1  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: dealii::FEValues<2, 2>::do_reinit()

#2  /home/dealii/build/lib/libdeal_II.g.so.9.3.0-pre: void dealii::FEValues<2, 2>::reinit<dealii::DoFHandler, false>(dealii::TriaIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2, 2>, false> > const&)

#3  /home/dealii/build/tests/quick_tests/step-slepc.debug: LaplaceEigenspectrumProblem::assemble_system()

#4  /home/dealii/build/tests/quick_tests/step-slepc.debug: LaplaceEigenspectrumProblem::run()

#5  /home/dealii/build/tests/quick_tests/step-slepc.debug: main

--------------------------------------------------------
```

**Comment by bangerth** [comment]
It would of course be very nice if I could report that that is the only test that fails, but it's really 310 tests among only the debug version tests :-) Clearly, that typo where I had accidentally said `if (n_threads > 0)` made the bug appear because it had tested cleanly before...

**Comment by bangerth** [comment]
Let's try this again. I think that what it boils down to is that if you say you're setting the value, you ought to be setting the value...

**Pull request merged**